### ISSUE TITLE
Vector subtraction in `look_at` inverts the forward vector

### ISSUE THREAD

**Issue by joncol** [bevyengine/bevy#1153]
**Bevy version**

0.4.0

**Operating system & version**

Linux

**What you did**

Calling `forward` on a `Transform` created by `look_at` gives the result in the wrong direction.

**What you expected to happen**

I was expecting `forward` called on a `Transform` (for a `Camera`) created by `look_at` to be pointing in the correct direction.

**What actually happened**

It was pointing backwards.

**Additional information**

The code of `look_at` looks like:
`let forward = Vec3::normalize(self.translation - target);`

This means the vector `forward` will go from `target` to `self.translation`, so in reality it will be pointing backward from the `target`, instead of forward...

I guess there must be the same bug in some projection code somewhere else for this not to be noticed before?

Or am I missing something obvious?

**Comment by DJMcNab** [comment]
Minimal example:
<details>
<summary>Code of the example</summary>

```rust
use bevy::prelude::*;

const USE_ORTHOGRAPHIC: bool = true;

fn main() {
    App::build()
        .add_plugins(DefaultPlugins)
        .add_startup_system(setup.system())
        .run();
}

/// set up a simple 3D scene
fn setup(
    commands: &mut Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<StandardMaterial>>,
) {
    let camera_transform = Transform::from_xyz(0., 0., 4.);
    assert_eq!(camera_transform.forward(), Vec3::unit_z());
    // add entities to the world
    commands
        // cube at 0,0
        .spawn(PbrBundle {
            mesh: meshes.add(Mesh::from(shape::Cube { size: 1.0 })),
            material: materials.add(Color::rgb(0.8, 0.7, 0.6).into()),
            ..Default::default()
        });

    if USE_ORTHOGRAPHIC {
        commands.spawn(OrthographicCameraBundle {
            transform: camera_transform,
            orthographic_projection: bevy::render::camera::OrthographicProjection {
                scaling_mode: bevy::render::camera::ScalingMode::FixedVertical,
                ..Default::default()
            },
            ..OrthographicCameraBundle::new_3d()
        });
    } else {
        commands.spawn(PerspectiveCameraBundle {
            transform: camera_transform,
            ..Default::default()
        });
    }
}

```
</details>

In this example, we create a camera with a transform in positive z, confirm that its forward direction is towards postiive z, and create a cube at the origin. However, the cube is still visible, even though it is behind the camera. This applies for both orthographic and perspective camera bundles

<details>
<summary> Screenshot of perspective bundle </summary>

![image](https://user-images.githubusercontent.com/36049421/106359984-0042c580-630e-11eb-9558-d008160bb61e.png)

</details>

This suggests that the camera is looking backwards, which seems peculiar

**Comment by DJMcNab** [comment]
Changing https://github.com/bevyengine/bevy/blob/57f9ac18d7ac039d94466c479b70802597569259/crates/bevy_render/src/camera/projection.rs#L22-L24 to use `perspective_rh` seemed to fix my minimal example. Reversing the `-` in `look_at also makes the other examples at least be looking in the right direction. However, this has signifant fallout in other examples (it makes all models render inside-out?).

However, looking at `alien_cake_addict` (added in https://github.com/bevyengine/bevy/pull/1252) a further effect of these changers is to invert the left and right arrows. 

Basically, this is all janky and strange and I don't like it

**Comment by alice-i-cecile** [comment]
@jamadazi @DJMcNab is this solved by #1361?

**Comment by DJMcNab** [comment]
No, that PR doesn't touch this code - that is the actual maths which is a problem is left untouched.

**Comment by Looooong** [comment]
From what I can see, the camera forward points toward -Z direction in local space. Therefore, we would expect that this method will rotate the camera to "look" at `target` correctly. However, if we use `look_at` for other cases, we will assume that it points the +Z direction toward `target` instead. Also, the method's document mentions the `local z direction`, not +Z or -Z direction. This is kinda confusing.

**Comment by superdump** [comment]
> However, if we use `look_at` for other cases, we will assume that it points the +Z direction toward `target` instead.

Why?

I was discussing this on Discord a couple of weeks ago when someone asked "what is forward?" Initially I was going to say that it is just -Z, but the answer became more complicated as I thought about it. It is kind of an implicit direction defined by the entity being transformed. For example, if you load a humanoid model where it looks along +Z in right-handed y-up, and give it an identity transform, then its 'forward' is along +Z. If it were looking along +X then forward for it would be along +X.

However, for a camera, because we use right-handed y-up perspective projections, that means that forward is explicitly along -Z. Maybe it is standard to consider forward as being out of the screen because that is how assets are modelled?

What is the expectation for the forward-facing direction of a model that is imported into a right-handed y-up application? I feel like it should face along the -Z axis and that this Transform::look_at implementation is wrong for a right-handed y-up world. But perhaps that is not the broader convention?

**Comment by superdump** [comment]
I asked @aclysma for input on this as I figured they may have insight. They provided this link to where glTF explicitly defines the coordinate space of the models as right-handed y-up **_with forward along +z_**! So that's one reference point. Do other model formats define what is forward as well? That would be great if so! :) Though it may mean that you need to convert from that format's definition of forward to whatever you want to use as your internal engine idea of forward at import-time. But as you may need to convert between coordinate spaces anyway, I guess that isn't a big problem.

**Comment by skorpi-and-friends** [comment]
I have to agree with @Looooong's comment that the documentation's confusing. Bevy seems to use [WebGPU's](https://gpuweb.github.io/gpuweb/#coordinate-systems) right-handed y-up coordinate system which will mean a choice between -Z as fwd, +X as right vs +Z fwd, -X right with the former being in use as evidenced by the current `look_at` implementation and the `forward` method on the Transform types. glTF uses the [-X is right](https://www.khronos.org/registry/glTF/specs/2.0/glTF-2.0.html#coordinate-system-and-units) convention but I think that'd be even more confusing for a game application. -Z fwd is at least familiar for people coming from Godot.

That being said, there's no bug in the `look_at` code as per the conventions bevy seems to be following now.

**Comment by superdump** [comment]
Revisiting this due to #5670 . I remember that while looking at (ha!) views for shadow mapping, I ended up using glam's look at matrix calculations because bevy's seemed to be the inverse, or vice versa. Anyway, looking at (ha!) the code we have in bevy, this is very confusing:
```rust
    pub fn look_at(&mut self, target: Vec3, up: Vec3) {
        let forward = Vec3::normalize(self.translation - target);
        let right = up.cross(forward).normalize();
        let up = forward.cross(right);
        self.rotation = Quat::from_mat3(&Mat3::from_cols(right, up, forward));
    }
```

Forward in bevy terms is -z. The cross-product in a right-handed coordinate system such as what we use in bevy works like this:
![Right-hand curling fingers around the shortest angle from vector A to vector B with the thumb pointing in the direction perpendicular to the two vectors](https://user-images.githubusercontent.com/302146/184522003-41733411-0aff-4838-af6b-7f60984c84d8.png)
(from https://physics.stackexchange.com/a/518085 ). So in the above code, forward is pointing from the target to this transform's translation which is backward, `up.cross(forward)` does indeed point to the right - if you curl the fingers of a right hand from up around to backward and stick the thumb out, it points to the right. And then `forward.cross(right)` is actually backward cross right which does indeed point up.

glam does the following:
```rust
    pub fn look_at_rh(eye: Vec3, center: Vec3, up: Vec3) -> Self {
        glam_assert!(up.is_normalized());
        Self::look_to_lh(eye, eye.sub(center), up)
    }
...
    fn look_to_lh(eye: Vec3, dir: Vec3, up: Vec3) -> Self {
        let f = dir.normalize();
        let s = up.cross(f).normalize();
        let u = f.cross(s);
        Self::from_cols(
            Vec4::new(s.x, u.x, f.x, 0.0),
            Vec4::new(s.y, u.y, f.y, 0.0),
            Vec4::new(s.z, u.z, f.z, 0.0),
            Vec4::new(-s.dot(eye), -u.dot(eye), -f.dot(eye), 1.0),
        )
    }
```
The translation component in the fourth column is also interesting but I'll ignore that for now.

glam would end up calling `look_to_lh` with our equivalent of `self.translation - target`. I think `f`, `s`, and `u` are forward, sideways, and up. They are calculated in the same way as bevy does. However, we use right, up, forward as the _columns_ of the matrix, but glam spreads them across the rows instead. Bevy is using the transpose of the 3x3 compared to glam. The transpose of a pure rotation matrix is its inverse. So that explains me observing that glam and bevy's `look_at` are inverses of each other - @cart .

I don't know why things appear to work properly in bevy given this.

**Comment by devil-ira** [comment]
```rust
pub fn look_at(&mut self, target: Vec3, up: Vec3) {
    // This calculates a vector going from the target to the transform, not the other way around.
    // It's correct because in Bevy the forward direction points to the _negative_ Z axis (ie. the opposite direction).
    let forward = Vec3::normalize(self.translation - target);
    // If we calculate the vector from the transform to the target instead Here's the same thing, written differently. Note the negation.
    let forward = -Vec3::normalize(target - self.translation);

    let right = up.cross(forward).normalize();
    let up = forward.cross(right);
    self.rotation = Quat::from_mat3(&Mat3::from_cols(right, up, forward));
}

// Here's the same method with different variable names.
pub fn look_at(&mut self, target: Vec3, up: Vec3) {
	// Forward is negative Z
    let neg_z = Vec3::normalize(self.translation - target);
	// Right is X
    let x = up.cross(neg_z).normalize();
    let y = neg_z.cross(x);
    self.rotation = Quat::from_mat3(&Mat3::from_cols(x, y, neg_z));
}
```

Hopefully that helped clear this up for someone.

As for what's up with `Mat4::look_at_rh`, I honestly don't know what it produces.
I'd at least expected the resulting translation of the matrix to be equal to `eye`, but it's not.
